python-spec/src/somabase/options.py

```diff
@@ -0,0 +1,95 @@
+"""Enums and other types used as options across methods of many types.
+
+These types are *concrete* and should be accepted as inputs
+"""
+
+import enum
+from typing import Optional
+
+import attrs
+
+
+@attrs.define(frozen=True)
+class IOfN:
+    """Specifies that a read should return partition ``i`` out of ``n`` total.
+
+    For a read operation that returns ``n`` partitions, the read operation will
+    return the ``i``th partition (zero-indexed) out of ``n`` partitions of
+    approximately equal size.
+    """
+    __slots__ = ()
+
+    i: int = attrs.field()
+    """Which partition to return (zero-indexed)."""
+    n: int = attrs.field()
+    """How many partitions there will be."""
+
+    @i.validator
+    def _validate(self, _, __):
+        del _, __  # Unused.
+        if not 0 <= self.i < self.n:
+            raise ValueError(
+                f"Partition index {self.i} must be in the range [0, {self.n})"
+            )
+
+
+@attrs.define(frozen=True)
+class BatchSize:
+    """Specifies the size of a batch that should be returned from reads.
+
+    Read operations on foundational types return an iterator over "batches" of
+    data, enabling processing of larger-than-core datasets. This class allows
+    you to control what the size of those batches is.
+
+    If none of these options are set, a "reasonable" batch size is determined
+    automatically.
+
+    For example::
+
+        BatchSize(count=100)
+        # Will return batches of 100 elements.
+
+        BatchSize(bytes=1024 ** 2)
+        # Will return batches of up to 1 MB.
+
+        BatchSize()
+        # Will return automatically-sized batches.
+    """
+    __slots__ = ()
+
+    count: Optional[int] = attrs.field(default=None)
+    """``arrow.Table``s with this number of rows will be returned."""
+    bytes: Optional[int] = attrs.field(default=None)
+    """Data of up to this size in bytes will be returned."""
+
+    @count.validator
+    @bytes.validator
+    def _validate(self, attr: attrs.Attribute, value):
+        if not value:
+            return  # None (or 0, which we treat equivalently) is always valid.
+        if value < 0:
+            raise ValueError(f"If set, '{attr.name}' must be positive")
+        if self.count and self.bytes:
+            raise ValueError("Either 'count' or 'bytes' may be set, not both")
+
+
+class BatchFormat(enum.Enum):
```

Comment:
I primarily defined these in the abstract spec because i was too lazy to define a read/write operation for every data type, and I just wanted to express which formats were expected to be used.

For the Python bindings, it seems more Python to use a specific method name for each type, rather than a generic read method that takes a batchformat parameter.  Should we pick one?

I.e., do we prefer:
> object.read(batch_format="table")

or
> object.read_as_arrow_table()

In the current NdArray prototype, I actually split the difference, and defined:
```
SOMASparseNdArray.read_sparse_tensor(format="coo|csr|csc")  -> pyarrow.SparseTensor subclass
SOMASparseNdArray.write_sparse_tensor(...)

SOMADenseNdArray.read_tensor() -> pyarrow.Tensor
SOMADenseNdArray.write_tensor(tensor)
SOMADenseNdArray.read_numpy() -> numpy.ndarray
SOMADenseNdArray.write_numpy(numpy.ndarray)

SOMASparseNdArray.read_table() -> pyarrow.Table # contains COO data in a Table
SOMASparseNdArray.read_as_pandas() -> pandas.DataFrame  # same as read_table
```

etc (you get the idea)

This seems to be more in keeping with Pandas/Numpy/Arrow API, and provides a clearer API IMHO.

Reply:
re
> object.read(batch_format="table")
vs.
> object.read_as_arrow_table()
i am personally happy either way

Reply:
OK, I’m glad to read this—I was going to ask about it more when we got to something that implemented that interface. I do think `read_as_x` has the advantage of being a little more common in the Python ecosystem, as well as less error-prone and more useful as far as type checking goes. I’m not a fan of functions which return different overall types depending upon a parameter you pass in; while it is possible to write annotations that cover it (e.g. with `@overload` and the `Literal[some_literal]` syntax), it still leaves the implementor (and potentially the user) with worse tooling feedback and the potential to miss or mismatch entries.

I’ll drop the `BatchFormat` enum for now (with the knowledge that, should we ever need it, it survives in changeset 49cbe30371d8343315c1c71192560452c62bea7f).

Reply:
me:

> i am personally happy either way

@thetorpedodog :

> I’m not a fan of functions which return different overall types depending upon a parameter you pass in; while it is possible to write annotations that cover it (e.g. with @overload and the Literal[some_literal] syntax), it still leaves the implementor (and potentially the user) with worse tooling feedback and the potential to miss or mismatch entries.

yes, this really tips it -- having specifier args on the input, and `Union[foo, bar]` on the return types -- does indeed muddy the keystroking as well as weakening the value of the type-checker.

so i officially throw in with @thetorpedodog and @bkmartinjr :D